perception/mil_vision/mil_vision_tools/rect_finder.py

```diff
@@ -0,0 +1,179 @@
+        '''
+        Creates a RectFinder from a geometry_msgs/Polygon message.
+
+        The length of the Rect becomes the range of x/y (whichever larger)
+
+        The width of the Rect becomes the range of x/y (whichever shorter)
+        '''
+        assert isinstance(polygon, Polygon)
+        arr = rosmsg_to_numpy(polygon.points)
+        # If it contains just one point, treat the x/y and length/width
+        if arr.shape[0] == 1:
+            return cls(arr[:,0][0], arr[:,1][0])
+        else:
+            return cls(np.ptp(arr[:,0]), np.ptp(arr[:,1]))
+
+    def to_polygon(self):
+        '''
+        Returns a geometry_msgs/Polygon representing the four corners
+        of the rectangle where all z = 0
+        '''
+        return numpy_to_polygon(self.model_3D)
+
+    @staticmethod
+    def sort_corners(rect, debug_image=None):
+        '''
+        Given a contour of 4 points, returns the same 4 points sorted in a known way.
+        Used so that indicies of contour line up to that in model for cv2.solvePnp
+
+        @param debug_image if not None, puts a circle and text of the index in each corner
+
+        p[0] = Top left corner         0--1     1--------2
+        p[1] = Top right corner        |  |  or |        |
+        p[2] = Bottom right corner     |  |     0--------3
+        p[3] = Bottom left corner      3--2
+
+        Credit to David Soto for this implementation.
+        '''
+        if rect.shape != (rect.shape[0], 2):
+            M = cv2.moments(rect)
+            rect = np.reshape(rect, (rect.shape[0], 2))
+        else:
+            temp = rect.reshape(rect.shape[0], 1, 2)
+            M = cv2.moments(temp)
+        centroid = np.array((int(M['m10']/M['m00']), int(M['m01']/M['m00'])))
+        vectors = rect - centroid
+        atan = np.arctan2( vectors[:,1], vectors[:,0] )
+        atan_indicies = np.argsort(atan)
+        # Sort by arctan formed by vector from centroid to each point
+        rect = rect[atan_indicies]
+        # If rect is horizontal, correct indicies as noted above
+        if np.linalg.norm(rect[0] - rect[1]) > np.linalg.norm(rect[0] - rect[3]):
+              rect = rect[[3, 0, 1, 2]]
+        # Print indicies onto image if debug_image is given
+        if debug_image is not None:
+            for i, pixel in enumerate(rect):
+                center = (int(pixel[0]), int(pixel[1]))
+                cv2.circle(debug_image, center, 5, (0, 255, 0), -1)
+                cv2.putText(debug_image, str(i), center, cv2.FONT_HERSHEY_SCRIPT_COMPLEX,1, (0,0,255))
+        return rect
+
+    def verify_contour(self, contour):
+        '''
+        Returns a numerical comparison between a contour and the perfect
+        model of the rectangle. A perfectly matched contour returns 0.0.
+
+        Useful for filtering contours.
+        '''
+        return cv2.matchShapes(contour, self.model_2D, 3, 0.0)
+
+    def get_corners(self, contour, epsilon_factor=0.02, debug_image=None):
```

Comment:
One way I reliably approximated contours as quadrilaterals was by doing it iteratively and increasing epsilon until I had a quadrilateral. Its more robust for when over-saturation or caustics make it hard to retrieve the quadrilateral. 

Check [this](https://github.com/uf-mil/SubjuGator/blob/5d169e9d23d6b20ef488040257fc0837ac887adb/perception/sub8_perception/nodes/torpedo_board.cpp#L1095) out for an implementation.